File: 2022/go/05.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"regexp"
	"strconv"
)

func main() {
	////// build the moves
	movesText, _ := os.Open("../data/05-moves.txt")	
	movesScanner := bufio.NewScanner(movesText)
	reMoves, _ := regexp.Compile("move (\\d+) from (\\d+) to (\\d+)")

	var movesList [][]string
	for movesScanner.Scan() {
		matches := reMoves.FindStringSubmatch(movesScanner.Text())
		matches = matches[1:]
		movesList = append(movesList, matches)
		//fmt.Println(matches)
	}

	////// build the stacks
	stacksText, _ := os.Open("../data/05-crates.txt")
	stacksScanner := bufio.NewScanner(stacksText)
	stacks := make([][]rune, 9)
	for stacksScanner.Scan() {
		for i, r := range stacksScanner.Text() {
			if r != ' ' && r!= '[' && r!= ']' && i%4 == 1 {
				stacks[i/4] = append(stacks[i/4], r)
			}
		}
	}
	// the stacks are in reverse order
	//fmt.Println(stacks)
	// initial state
	for _, stack := range stacks {
		fmt.Printf("%c", stack[0])
	}
	fmt.Println("")
	part2Stacks := make([][]rune, 9)
	copy(part2Stacks, stacks)
	//fmt.Println(part2Stacks)

	for i := 0; i < len(movesList); i++ {
		moves := movesList[i]
		count, _ := strconv.Atoi(moves[0])
		from, _ := strconv.Atoi(moves[1])
		from -= 1
		to, _ := strconv.Atoi(moves[2])
		to -= 1
		for i := 0; i < count; i++ {
			//// part 1 work
			// add to the 'to'
			stacks[to] = append([]rune{stacks[from][0]}, stacks[to]...)
			// take away from the 'from'
			stacks[from] = stacks[from][1:]

			//// part 2 work
			// similar to part 1, still add them 1 at a time, but in the reverse order
			part2Stacks[to] = append([]rune{part2Stacks[from][(count-1)-i]}, part2Stacks[to]...)
		}
		// only do this once per move
		part2Stacks[from] = part2Stacks[from][count:]
	}
	// part 1 answer
	for _, stack := range stacks {
		fmt.Printf("%c", stack[0])
	}

	fmt.Println("")

	// part 2 answer
	for _, stack := range part2Stacks {
		fmt.Printf("%c", stack[0])
	}
}

File: 2022/python/05.py
import re
from copy import deepcopy

movesText = open("../data/05-moves.txt").read().split("\n")

moveRe = re.compile(r"move (\d+) from (\d+) to (\d+)")
moves = [re.findall(moveRe, move)[0] for move in movesText]
#print(moves) # this is a list of tuples

def part1(stacks, moves):
  for move in moves:
    for _ in range(int(move[0])):
      stacks[int(move[2])-1].append(stacks[int(move[1])-1].pop())
  return stacks

def part2(stacks, moves):
  for move in moves:
    stacks[int(move[2])-1] += stacks[int(move[1])-1][-int(move[0]):]
    stacks[int(move[1])-1] = stacks[int(move[1])-1][:-int(move[0])]
  return stacks

# initialize the stacks, and fill them with the crates at time zero
"""
I found the easiest way to do this was to remove the legend from the source data 
and replace whitespace with '_'. After you do this, you can reverse the list 
and it will be in the right order.
"""
stacks = [[] for _ in range(9)]
cratesText = open("../data/05-crates.txt").read().replace(" ", "_").split("\n")
cratesText.reverse()
# not pythonic, but it gets the job done :shrug:
for line in cratesText:
  for i in range(1, len(line), 4):
    crate = line[i]
    if crate != "_":
      stacks[i//4].append(crate)
#print(stacks)
"""
deepcopy is necessary because otherwise the list of lists will 
not be a true copy, they'll both reference the same 
space in memory and overwrite one another
"""
p1Stacks = deepcopy(stacks)
p2Stacks = deepcopy(stacks)

endState1 = part1(p1Stacks, moves)
endState2 = part2(p2Stacks, moves)

# initial state
print("".join([stack[-1] for stack in stacks]))
# part 1 answer
print("".join([stack[-1] for stack in endState1]))
# part 2 answer
print("".join([stack[-1] for stack in endState2]))
